### PR TITLE
Change check of test method to looking for Xunit.FactAttribute attribute

### PR DESCRIPTION
**Pull request by genlu**
FYI @dotnet/roslyn-interactive @dotnet/roslyn-compiler

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
Does this cover derived types of Fact?

File: src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs [hunk 1]
```diff
@@ -577,7 +594,65 @@ static void TestMain()
...
+
+    [Xunit.Fact]
+    static int MethodWithFactAttribute(int x)
+    {
+        var y = x + 1;
+        var z = y + 1;
+        return z;
+    }
+
+    static int MethodWithOutFactAttribute(int x)
```
Comment (JohnHamby) [inline src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs]: It takes some cleverness to be able to tell from the test output which method is doing the flushing. Consider checking the generated IL (see an example above) or calling the method that does the flushing twice.

**Comment by genlu** [comment]
No, I made this change just so we don't have to change test methods for every projects we are testing against. The proper fix for the test identification is still tracked by #9811.

File: src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs (hunk 1 above)
Comment (genlu) [inline src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs]: OK, I've added IL verification.

**Comment by genlu** [comment]
@jcouv Could you please take a look at this? This is not intended to be the proper fix for the problem (which is still being tracked by the original issue), we just changed it here to simplify the process of our development.

**Comment by jcouv** [comment]
@genlu. Thanks for the ping. I'll look now.

**Comment by jcouv** [comment]
LGTM. 
I am curious about the long-term design for identifying test methods. Will that logic be hard-coded into the compiler (to understand the test frameworks we know about), or will it be pluggable by the user?

**Comment by JohnHamby** [comment]
The long-term--even the medium-term--design is for the compiler to get out of the business of identifying test methods and have that provided by the mechanism that runs tests. @genlu has improved a temporary expedient, but it's still temporary.

**Comment by JohnHamby** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs [hunk 2]
```diff
@@ -128,7 +128,15 @@ private static MethodSymbol GetFlushPayload(CSharpCompilation compilation, CShar
         private static bool IsTestMethod(MethodSymbol method)
         {
             // PROTOTYPE (https://github.com/dotnet/roslyn/issues/9811): Make this real. 
-            return method.Name.StartsWith("Test");
+            var attributes = method.GetAttributes();
+            foreach (var attribute in attributes)
+            {
+                if (attribute.IsTargetAttribute("Xunit", "FactAttribute"))
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs]: As noted in the conversation this is a prototype line.  We should reference the actual bug tracking the work here:

https://github.com/dotnet/roslyn/issues/9811

**Comment by jaredpar** [comment]
Like @jcouv and @jasonmalinowski I'm also curious about the long term plan for identifying test methods.  We certainly can't be hard coding knowledge of xunit into the compiler, or any other test farmework.

**Comment by KevinH-MS** [comment]
@jaredpar you're right.  Long term, this needs to be a function of the test driver, not the compiler.  I believe we have plans to make the test driver poke at the assembly under test to tell it to flush its coverage data.  We just haven't had a chance to make that happen yet.  This is a stop gap for a demo we're trying to put together.  @genlu can perhaps comment with more detail.

**Comment by jaredpar** [comment]
@KevinH-MS if that's the case we should be very explicit about this in the code.  It should be marked as PROTOTYPE in code and tracked with a bug number.

**Comment by KevinH-MS** [comment]
@jaredpar I agree.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by genlu** [comment]
@jcouv @jaredpar Yes, I agree this should not be in the compiler. As @KevinH-MS and @JohnHamby mentioned, this is just a quick fix I made for our demo next week. The long term plan for this is that compiler only needs to instrument code without doing anything special for test method, and it is test driver's responsibility to identify tests and call flushing method (which means, the code in compiler that identifies test method and wraps the method body with try-finally will be removed). Actually it won't be that long, we probably can achieve this in a week or so, I already have related stuff implemented on test-impact side, but it's not ready to be merged yet (the compiler change will come with it).

**Comment by genlu** [comment]
@jaredpar We have the issue link in the comment above the change.

**Comment by jaredpar** [comment]
@genlu ah okay.  That is the first line that github clipped hence it wasn't visible in my view.

**Pull request merged**